File: framework/src/gmsec4/util/Condition.cpp
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 *  @file Condition.cpp
 *
 *  @brief This file contains a class for cross platform condition variable support
 */

#ifndef WIN32
#include <errno.h>
#endif

#ifdef GMSEC_VC6
#define _WIN32_WINNT 0x0400
#endif

#include <gmsec4/util/Condition.h>

#include <gmsec4/internal/util/InternalCondition.h>


namespace gmsec {
namespace api {
namespace util {

Condition::Condition(Mutex& m)
	: m_internal(new gmsec::api::internal::InternalCondition(m))
{
}


Condition::Condition()
	: m_internal(new gmsec::api::internal::InternalCondition())
{
}


Condition::~Condition()
{
	delete m_internal;
}


Mutex& Condition::getMutex()
{
	return m_internal->getMutex();
}


int Condition::wait()
{
	return m_internal->wait();
}


int Condition::wait(long timeout)
{
	return m_internal->wait(timeout);
}


int Condition::waitUntil(const GMSEC_TimeSpec& until)
{
	return m_internal->waitUntil(until);
}


void Condition::signal(int result)
{
	m_internal->signal(result);
}


void Condition::broadcast(int result)
{
	m_internal->broadcast(result);
}

} // namespace util
} // namespace api
} // namespace gmsec

File: csharp/gmsec4/dox/util/Condition.cs
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */

/* This file is used for generating Doxygen documentation; it only contains
   stubs of constructors and methods associated with the class.
*/


namespace GMSEC.API {

/// <summary>%Condition support for managing dependent threads.</summary>

public class Condition : global::System.IDisposable {

  /// <summary>Standard constructor</summary>
  public Condition() {
  }

  /// <summary>Standard constructor that initializes with an externally managed mutex</summary>
  /// <param name="mutex">External Mutex object</param>
  public Condition(Mutex mutex) {
  }

  /// <summary>Returns a reference to the Mutex held by the object</summary>
  public Mutex GetMutex() {
  }

  /// <summary>Will block until the thread is signaled</summary>
  public int Wait() {
  }

  /// <summary>
  /// Will block until the thread is signaled or the time specified has expired. If millis is not
  /// positive, timeout immediately.
  /// </summary>
  ///
  /// <param name="millis">The length in time in milliseconds to wait for the thread to be signaled before continuing</param>
  public int Wait(int millis) {
  }

  /// <summary>
  /// The function will cause the calling thread to wait until the condition is signalled or the time specified has been reached
  /// </summary>
  ///
  /// <param name="t">The absolute timeout</param>
  public int WaitUntil(GMSEC_TimeSpec t) {
  }

  /// <summary>Signals any waiting threads to continue with their activity</summary>
  ///
  /// <param name="reason">The signal to emit to threads that are waiting</param>
  public void Signal(int reason) {
  }

  /// <summary>Signals all waiting threads to continue with their activity</summary>
  ///
  /// <param name="reason">The signal to emit to threads that are waiting</param>
  public void Broadcast(int reason) {
  }

  public static readonly int INVALID = 0;
  public static readonly int TIMEOUT = 1;
  public static readonly int USER    = 2;

}

}
